### ISSUE TITLE
Plugin incorrectly(?) reporting unused dependency

### ISSUE THREAD

**Issue by Mahoney** [autonomousapps/dependency-analysis-android-gradle-plugin#298]
**Build scan link**
https://scans.gradle.com/s/wqljzkve3gazi

**Plugin version**
0.61.0

**Gradle version**
gradle-6.8-branch-eskatos_kotlin_embedded_1_4-20200922125722+0000

**Android Gradle Plugin (AGP) version**
N/A

**Describe the bug**
Getting a failure claiming I have an unused dependency that I think is invalid - possibly due to the plugin not knowing about the `testFixtures` configuration? Or not knowing about how kotlin works with top level functions?

**To Reproduce**
1. Check out git@github.com:Mahoney-playground/goos.git
2. Run ./gradlew build

**Expected behavior**
I think the build should pass

**Additional context**
https://docs.gradle.org/current/userguide/java_testing.html#sec:java_test_fixtures

**Comment by autonomousapps** [comment]
You might be right about test fixtures. This plugin currently makes no accommodation for that.

You can get some insight into why the plugin is recommending something by eg:
```
./gradlew module:reason --id <identifier>
```
I'd give you an example for `<identifier>`, but you didn't tell me which dependency had been incorrectly declared unused :)

**Comment by Mahoney** [comment]
Yes, sorry, that was a rubbish bug report - rushing to get to bed. Added the obvious extra details.

**Comment by Mahoney** [comment]
I realise I've muddied the waters by using a snapshot version of gradle, but I see exactly the same behaviour reverting to gradle 6.6.1.

**Comment by Mahoney** [comment]
Updated the bug report to a branch using gradle 6.6.1 - same behaviour.

**Comment by autonomousapps** [comment]
Thanks, appreciated. I think this must be due to the plugin having no support yet for `testFixtures`. I will update this ticket to be a "feature request" to support that gradle feature.

For now, I think you can configure the plugin to ignore this "unused" dependency, following guidance in the wiki at https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/wiki/Customizing-plugin-behavior

**Comment by daanschipper** [comment]
I've done some more investigation into this issue. The `locateDependenciesMain` task does not take the capabilities of a project dependency into account when determining all declared dependencies.

Some information about a dependency is lost when iterating over all the dependencies and mapping it to a set in [ConfigurationsToDependenciesTransformer](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/eafc9b53c5ac1c2b7aa28172290039a10b378810/src/main/kotlin/com/autonomousapps/internal/ConfigurationsToDependenciesTransformer.kt#L47), only the identifier and configuration of a dependency is saved.

This causes issues when having dependencies declared as follows
```
project(':a') {
  plugins {
    id 'java-library'
    id 'java-test-fixtures'
  }
}

project(':b') {
  plugins {
     id 'java-library'
  }
  dependencies {
    testImplementation project(":a")    
    testImplementation testFixtures(project(":a"))
  }
}
```

only one 'location' is saved.
```
  {
    "identifier": ":a",
    "configurationName": "testImplementation",
    "isInteresting": true
  },
```

What is interesting, `artifactsReportMain` is able to locate the correct physical jars, but again the capability information is lost. This causes confusing reports down the line, e.g. the initial issue report.

```
  {
    "dependency": {
      "identifier": ":a",
      "configurationName": "testImplementation"
    },
    "isTransitive": false,
    "file": "<workspace>/a/.build/libs/al.jar"
  },
    {
    "dependency": {
      "identifier": ":a",
      "configurationName": "testImplementation"
    },
    "isTransitive": false,
    "file": "<workspace>/a/.build/libs/a-test-fixtures.jar"
  },
```

A potential solution would be to extend the `Location` data class with an additional field for the requested capabilities. By default this would be an empty list, but in case of `testFixtures` it would be (as you might expect) a list with one element `testFixtures`. Before diving further into this and start implementing the change, what do you think about this?

**Comment by autonomousapps** [comment]
@daanschipper thank you for the detailed note. I have given it some thought and I think you're on the right track, at least to the extent that it certainly seems like the `Location` class needs to encapsulate more information. That would only be a first step, of course. If you want to pursue this idea, feel free to file a PR and we can discuss further.

**Comment by autonomousapps** [comment]
FYI I am currently working on a massive redesign of the plugin, and as part of that I'm reworking the "location" code, and it will include an `attributes: Set<String>` (or something similar) to capture this information.

**Comment by autonomousapps** [comment]
Hey @jjohannes any thoughts on how we might support testFixtures?

**Comment by jjohannes** [comment]
@autonomousapps I would have to look into the code again for concrete implementation suggestions (which I can do soonish, if you plan to work on this).

General thoughts:

I would attempt to support arbitrary **additional source sets** and **feature variants**. Test fixtures uses this and has nothing 'special' on top from the dependency analysis perspective.

Then there are two aspects:
1. Dependencies **from** an additional source set - e.g. `testFixtureImplementation("commons-io:commons-io:2.6")`
2. Dependencies **to** an additional feature variant - e.g. `implementation(testFixtures(project(":module-a")))`

Note that `implementation(testFixtures(project(":module-a")))` is just a shortcut for 

```
implementation(project(":module-a")) {
  capabilities { requireCapability("my.group:module-a-test-fixtures") }
}
```

which would be the notation you can use for any _feature variant_. (For those following this issue and are interested in more details, I explained some of it here: https://youtu.be/XCzyUESaBHQ)

It might probably be a good idea to implement these two aspected one after the other. I _think_ that (1) is easier to implement, because there we do not need more information about which aritfacts (Jars/classes) to select on the "other side". We continue to look at the "main" Jar(s) only. Solving this, would already be a huge benefit. Not only for dependencies you declare _from_ `testFixtures`, but also from other additional source sets (like `integrationTest`).

For (2) we would need to make the connection between the capability you select in the dependency declaration and the aritfact(s) that resolves to on the "other side". Which in the test fixture example would not be `module-a.jar`, but `module-a-test-fixtures.jar`.

I once added these test cases that supposedly make sure that both cases do not lead to _wrong_ suggestions. They, and the changes made when they were introduced, could be a good starting point: https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/main/src/functionalTest/groovy/com/autonomousapps/jvm/CustomSourceSetSpec.groovy

**Comment by jjohannes** [comment]
Additional note: I am not sure if everything I wrote also applies for the Android `testFixtures`. I haven't looked at how that's done in AGP yet.

**Comment by autonomousapps** [comment]
This issue feels like the largest remaining gap in what this plugin can do. So, I think it's a good candidate as a next feature. That said, I'm extremely busy at work and don't know when I'll have time to dig into it.

**Comment by jjohannes** [comment]
> This issue feels like the largest remaining gap in what this plugin can do. 

I agree and I would also love to contribute. If I find some time, I'll have a closer look and see if I can get started with something.
